src/Features/CSharp/Portable/CodeFixes/AddExplicitCast/AddExplicitCastCodeFixProvider.cs

```diff
@@ -0,0 +1,372 @@
+                    if (IsArgumentListAndParameterListPerfactMatch(semanticModel, argumentList.Arguments, methodSymbol.Parameters, targetArgument, cancellationToken, out var paramIndex))
+                    {
+                        var correspondingParameter = methodSymbol.Parameters[paramIndex];
+                        var argumentConversionType = correspondingParameter.Type;
+
+                        if (correspondingParameter.IsParams && correspondingParameter.Type is IArrayTypeSymbol arrayType && !(targetNodeType is IArrayTypeSymbol))
+                        {
+                            // target argument is matched to the parameter with keyword params
+                            argumentConversionType = arrayType.ElementType;
+                        }
+
+                        mutablePotentialConversionTypes.Add(argumentConversionType);
+                    }
+                }
+
+                // Sort the potential conversion types by inheritance distance
+                var comparer = new InheritanceDistanceComparer(semanticModel, targetNodeType);
+                mutablePotentialConversionTypes.Sort(comparer);
+                mutablePotentialConversionTypes = mutablePotentialConversionTypes.Distinct().ToList(); // clear up duplicate types
+            }
+
+            // For cases like object creation expression. for example:
+            // Derived d = [||]new Base();
+            // It is always invalid except the target node has explicit conversion operator.
+            // filter function is going to filter the cases that are invalid
+            var nodeType = targetNodeType;
+            Func<ITypeSymbol, bool> filter = (targetNodeConversionType =>
+            {
+                var commonConversion = semanticModel.Compilation.ClassifyCommonConversion(nodeType, targetNodeConversionType);
+                if (targetNode.IsKind(SyntaxKind.ObjectCreationExpression) && !commonConversion.IsUserDefined)
+                {
+                    return false;
+                }
+                return commonConversion.Exists;
+            });
+
+            potentialConversionTypes = mutablePotentialConversionTypes.Where(filter).ToImmutableArray();
+            return !potentialConversionTypes.IsEmpty;
+        }
+
+        /// <summary>
+        /// Try to test if the invocation node is available to invoke the method.
+        /// </summary>
+        /// <param name="semanticModel"></param>
+        /// <param name="arguments">The arguments of invocation</param>
+        /// <param name="parameters">The parameters of </param>
+        /// <param name="targetArgument">The target argument that contains target node</param>
+        /// <param name="cancellationToken"></param>
+        /// <param name="targetParamIndex">Output the corresponding parameter index of the target arugment</param>
+        /// <returns>
+        /// True, if arguments and parameters match perfectly.
```

Comment:
if arguments/parameters match perfectly... then why would this be a 'candidate symbol' instead of just 'the symbol that we properly bound to'?

What's a code case where this feature is offered and this returns true?

Reply:
The function returns true if the arguments types can be converted to the parameters types.
The example is like this:
```csharp
class Base { }
class Derived1 : Base { }
class Derived2 : Base { }
class Derived3 : Base { }
void DoSomething(int i, Derived1 d) { }
void DoSomething(string s, Derived2 d) { }
void DoSomething(int i, Derived3 d) { }

Base b;
DoSomething(1, [||]b);
```

Since there is no `DoSomething(int i, Base b)` that matches `DoSomething(1, b)` exactly, `Symbol` is `null`.
If `b` can convert to other types, the invocation expression might be valid, so we look into `CandidateSymbols`. `CandidateSymbols` has `void DoSomething(int i, Derived1 d), void DoSomething(string s, Derived2 d), void DoSomething(int i, Derived3 d)`. After we compare the function arguments and parameters, we figure out that only `void DoSomething(string s, Derived2 d)` is not the perfect match candidate function for `DoSomething(1,b)` because int cannot convert to string. Thus, `Derived1` and `Derived3` are potential fixes for `b`.

Reply:
I think the wording here is potentially the problem. "Perfect" match means no binding error would happen (and the diagnostic wouldn't exist). This is more of a "CanArgumentListMatchParameterList" right?

Reply:
Yup, I am also thinking about wording, this name may mislead people. The method tests if all argument types are **convertible** to their corresponding parameter types, what about "CanArgumentTypesConvertToParameterTypes"?

Reply:
> The function returns true if the argument types can be converted to the corresponding parameter types.
The example is like this:

Great sentence.  Based on that:  `CanArgumentTypesBeConvertedToParameterTypes(...)`.